### PR TITLE
[LibraAccount] Add capability for key rotation

### PR DESCRIPTION
**Pull request by sblackshear**
This adds a new capability to LibraAccount that allows the account holder to delegate the privilege to rotate the account's authentication key. The mechanics are identical to the withdrawal capability added in 850513f194e40f1a37fd714a28db469de7204cce.

This feature enables several interesting use-cases:
- Warm wallets: A wallet at address A wants to spend frequently by signing transactions with the auth key for A, which increases the likelihood that the key could be compromised. But if the wallet delegates its key rotation capability to another account whose key is colder, this gives the owner recourse in the case that the auth key is compromised. The capability can be used to reset the auth key to something new.
- Account recovery: An account owner can delegate the key rotation capability elsewhere and bury the private key (similar to the "recovery codes" mechanism implemented for many web services). This allows the owner to recover usage of their account even if they lose their primary private key.
- Partially custodial wallets: an account owner can allow a custodial provider to manage the private key they use for spending, but delegate their key rotation capability to an address that only the owner knows the key for. That allows the owner to reclaim their account from the custodian if they need or want to.

## Motivation

Supporting our custody team by providing the features they need.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Many times.

## Test Plan

New E2E tests.

### CODE REVIEW

**Comment by davidiw** [comment]
The build error seems related to having a Rust, Move, and VM definitions of data structures.

File: language/stdlib/modules/libra_account.mvir [hunk 1]
```diff
@@ -14,6 +14,9 @@ module LibraAccount {
         authentication_key: bytearray,
         // The coins stored in this account
         balance: LibraCoin.T,
+        // If true, the authority to rotate the authentication key of this account resides elsewhere
```
Comment (weilei0107) [inline language/stdlib/modules/libra_account.mvir]: Would it be possible to preserve the ability to rotate auth key also for the account owner? That way we could still do the periodic proactive self key rotation using the hot key itself but also have a backup mechanism to rotate in the delegated account in case the hot key is compromised and rotated by malicious actors.

File: language/functional_tests/tests/testsuite/libra_account/delegate_rotation_capability.mvir [hunk 2]
```diff
@@ -0,0 +1,115 @@
+//! account: alice
+//! account: bob
+
+//! sender: alice
+// A module that lets the account owner rotate their auth key as they normally would, but also
+// allows a different address `master_key_address` to rotate the auth key. This is useful for
+// implementing (e.g.) a warm wallet with a cold recovery key.
```
Comment (kchalkias) [inline language/functional_tests/tests/testsuite/libra_account/delegate_rotation_capability.mvir]: Better use the word hot instead of warm `(e.g., a hot wallet with a cold recovery key)`.

File: language/stdlib/modules/libra_account.mvir (hunk 1 above)
Comment (kchalkias) [reply language/stdlib/modules/libra_account.mvir]: I agree, both the auth_key and cold_rotation_key should be able to rotate the `auth_key`.

I guess when used in practice some implementations would allow optionality on selecting the desired policy, but let's better keep it simple for this example.

Also, if one uses both capabilities, we should mention somewhere that controlling the rotation capability does not necessarily give you spending power. Good implementations should be carefully designed to  avoid such edge-cases, i.e., by including both capabilities in the module logic.

Can we also add a comment or test on what happens if there is a rotation cycle A->B->A or A->B->C->A?

**Comment by sblackshear** [comment]
@davidiw 
> Is the cost of having a capability is associated with the account owner?

Do you mean the cost of storing the capability in global storage, or something else (assuming the former)? It depends on where you store the capability. If you store it in a resource under the original owner's account (as we do in the `SharedRotation` module in the tests), the owner will pay. If you store it under a different account (which the owner of that account will need to consent to), that account will pay

>How does a single account handle multiple capabilities given to it? Is it as trivial as using a Vector instead of a single value?
 
The account can design whatever scheme it likes to manage multiple capabilities; a Vector would indeed be one good way.

>Can the holder of a capability transfer it to another account?

It depends on how the capability is stored. In the `SharedRotation` example in the tests, the capability gets "locked" in the `SharedRotation.T` resource and thus can never be transferred elsewhere. But we could have just as easily written the code in a way that allowed it to be transferred back to the original owner or to someone else.

>How does one identify the holder of a capability?
There's two ways one might think of "holder". One is unambiguous, but perhaps not-so-useful: the account address that the resource holding the capability is stored under. For example, in the `SharedRotation` code, the original account owner stores the `SharedRotation.T` resource.

The other definition of "holder" is more ambiguous, but more useful: who can acquire a reference to the capability (and thus use it to rotate the key)? In the `SharedRotation` example, we have written the code so that both the original account holder Alice and the delegate Bob are "holders" in this sense. 

> I think you had mentioned a few resources that motivated you to go in this direction versus the account having maintaining some sort of ACL-based access via smart contracts. Could you share some pointers?

[This](http://srl.cs.jhu.edu/pubs/SRL2003-02.pdf) paper is great. They use this analogy of row-based vs column-based authority to explain the difference between capabilities and ACL and discuss several advantages of the capability-based approach over ACLs. 

>On another note, this changes the definition of an account. Once we get to mainnet, do we have a proposal on how we would handle changes like these? I imagine the account could store the resource but it would not necessarily be defined as a resource in the account but more associated with the address of the account? LIke move_to_sender? Maybe? Still trying to wrap my head around move :).

Great question. In general, these sorts of breaking change will be much harder/more ill-advised once we get to mainnet. But there are design patterns and tricks (such as the one you mention: publish the resource to add as a separate resource via `move_to_sender` instead of actually changing the `LibraAccount.T` resource as this PR does) that can be used, particularly if the module is designed with upgradability in mind. @sunshowers has done a lot of thinking about this; I will send you some links to relevant docs/discussions.

File: language/stdlib/modules/libra_account.mvir (hunk 1 above)
Comment (sblackshear) [reply language/stdlib/modules/libra_account.mvir]: The use-case @weilei0107 has in mind can indeed be supported with the existing code (and I have changed the E2E test to demonstrate this).

Good point about the odd cases that can come up with using both the withdrawal and rotation capabilities + cycles. I'd like to spend some more time thinking about/documenting the best practices + writing tests after getting this change in. I think this will fit nicely with trying to write formal specifications for our hot/cold wallets (@DavidLDill will be excited).

File: language/stdlib/modules/libra_account.mvir [hunk 3]
```diff
@@ -196,12 +207,71 @@ module LibraAccount {
...
     // Rotate the transaction sender's authentication key
     // The new key will be used for signing future transactions
     public rotate_authentication_key(new_authentication_key: bytearray) acquires T {
         let sender_account: &mut Self.T;
+
         sender_account = borrow_global_mut<T>(get_txn_sender());
-        *(&mut move(sender_account).authentication_key) = move(new_authentication_key);
+        if (*&copy(sender_account).delegated_key_rotation_capability) {
+            // The sender has delegated the privilege to rotate her key elsewhere--abort
+            abort(11);
```
[the remark this answers is not in the document]
Comment (sblackshear) [reply language/stdlib/modules/libra_account.mvir]: Good point; we need to do that in a follow-up. I don't want to add it to errors.rs right now because it will just end up being an unused constant, but we should document all of these soon.
